File: backend/schemas/compensations.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

from models.compensations import (
    CompensationLogAction,
    CompensationStatus,
    CompensationType,
    DS_ANNUAL_REFERENCE_QUOTA,
    UserCompensationStatus,
)


class CompensationEventCreate(BaseModel):
    event_type: CompensationType
    motivo: str = Field(..., min_length=3, max_length=8000)
    participant_user_ids: list[int] = Field(..., min_length=1)


class CompensationEventUpdate(BaseModel):
    event_type: CompensationType | None = None
    motivo: str | None = Field(default=None, min_length=3, max_length=8000)
    participant_user_ids: list[int] | None = Field(default=None, min_length=1)


class CompensationDecisionBody(BaseModel):
    motivo: str | None = Field(default=None, max_length=4000)


class CompensationRejectBody(BaseModel):
    motivo: str = Field(..., min_length=1, max_length=4000)


class CompensationActionBody(BaseModel):
    motivo: str = Field(..., min_length=1, max_length=4000)


class CompensationEventPublic(BaseModel):
    id: int
    event_type: CompensationType
    motivo: str
    status: CompensationStatus
    created_by_id: int
    decided_by_id: int | None
    decided_at: datetime | None
    decision_motivo: str | None
    created_at: datetime
    updated_at: datetime
    participant_user_ids: list[int] = Field(default_factory=list)
    created_by_label: str | None = None
    decided_by_label: str | None = None

    model_config = {"from_attributes": False}


class CompensationEventLogPublic(BaseModel):
    id: int
    compensation_event_id: int
    actor_id: int
    actor_label: str
    action: CompensationLogAction
    from_status: CompensationStatus | None
    to_status: CompensationStatus | None
    motivo: str | None
    details: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class UserCompensationPublic(BaseModel):
    id: int
    user_id: int
    compensation_event_id: int
    status: UserCompensationStatus
    created_at: datetime
    display_label: str = ""

    model_config = {"from_attributes": True}


class UserCompensationAvailablePublic(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: int
    compensation_type: CompensationType = Field(validation_alias="type", serialization_alias="type")
    label: str
    event_date: date
    description: str


class DsUsagePublic(BaseModel):
    user_id: int
    year: int
    used_count: int
    reference_quota: int = DS_ANNUAL_REFERENCE_QUOTA
    display: str


class CompensationDashboardSummary(BaseModel):
    pending_count: int
    approved_recent_count: int
    ds_usage_samples: list[DsUsagePublic]
    recent_events: list[CompensationEventPublic]

File: frontend/src/types/compensations.ts
export type CompensationType =
  | "CPJ_SUPPORT"
  | "WEAPON_OCCURRENCE"
  | "RELEVANT_OCCURRENCE"
  | "TWO_WANTED"
  | "FIVE_FLAGRANTS"
  | "FOLGA_MENSAL"
  | "COMPENSACAO"
  | "DS";

export type CompensationEventStatus =
  | "PENDING"
  | "APPROVED"
  | "REJECTED"
  | "CANCELLED"
  | "REVERTED";

export type CompensationLogAction =
  | "CREATED"
  | "APPROVED"
  | "REJECTED"
  | "UPDATED"
  | "CANCELLED"
  | "REVERTED";

export const COMPENSATION_TYPE_LABELS: Record<CompensationType, string> = {
  CPJ_SUPPORT: "Apoio CPJ / operacional (≥4h)",
  WEAPON_OCCURRENCE: "Ocorrência com armas",
  RELEVANT_OCCURRENCE: "Ocorrência de grande relevância (N90/TAT)",
  TWO_WANTED: "02 procurados",
  FIVE_FLAGRANTS: "05 flagrantes",
  FOLGA_MENSAL: "Folga mensal",
  COMPENSACAO: "Compensação",
  DS: "Dispensa de serviço (DS)",
};

export const OPERATIONAL_COMPENSATION_TYPES: CompensationType[] = [
  "FOLGA_MENSAL",
  "COMPENSACAO",
  "DS",
];

export const MERIT_COMPENSATION_TYPES: CompensationType[] = [
  "CPJ_SUPPORT",
  "WEAPON_OCCURRENCE",
  "RELEVANT_OCCURRENCE",
  "TWO_WANTED",
  "FIVE_FLAGRANTS",
];

export const COMPENSATION_STATUS_LABELS: Record<CompensationEventStatus, string> = {
  PENDING: "Pendente",
  APPROVED: "Aprovado",
  REJECTED: "Indeferido",
  CANCELLED: "Cancelado",
  REVERTED: "Revertido",
};

export interface CompensationEventPublic {
  id: number;
  event_type: CompensationType;
  motivo: string;
  status: CompensationEventStatus;
  created_by_id: number;
  decided_by_id: number | null;
  decided_at: string | null;
  decision_motivo: string | null;
  created_at: string;
  updated_at: string;
  participant_user_ids: number[];
  created_by_label?: string | null;
  decided_by_label?: string | null;
}

export interface CompensationEventLogPublic {
  id: number;
  compensation_event_id: number;
  actor_id: number;
  actor_label: string;
  action: CompensationLogAction;
  from_status: CompensationEventStatus | null;
  to_status: CompensationEventStatus | null;
  motivo: string | null;
  details: string | null;
  created_at: string;
}

export interface DsUsagePublic {
  user_id: number;
  year: number;
  used_count: number;
  reference_quota: number;
  display: string;
}

export interface CompensationDashboardSummary {
  pending_count: number;
  approved_recent_count: number;
  ds_usage_samples: DsUsagePublic[];
  recent_events: CompensationEventPublic[];
}

export interface CompensationEventCreatePayload {
  event_type: CompensationType;
  motivo: string;
  participant_user_ids: number[];
}

export interface CompensationEventUpdatePayload {
  event_type?: CompensationType;
  motivo?: string;
  participant_user_ids?: number[];
}
